File: domains/knowledge-update/check_capabilities_sync.py
#!/usr/bin/env python3
"""Check that DataFinder manifest endpoint capabilities are represented."""

from __future__ import annotations

import json
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
MANIFEST = ROOT / "domains" / "datafinder-interface" / "manifest.json"
CAPABILITIES = ROOT / "domains" / "intent-routing" / "capabilities.json"


def _load(path: Path) -> dict:
    return json.loads(path.read_text(encoding="utf-8"))


def main() -> int:
    manifest = _load(MANIFEST)
    capabilities = _load(CAPABILITIES)
    manifest_ids = {
        ep["capability_id"]
        for ep in manifest.get("endpoints", [])
        if ep.get("capability_id")
    }
    capability_ids = {
        item["capability_id"]
        for item in capabilities.get("capabilities", [])
        if item.get("capability_id", "").startswith("datafinder.openapi.")
    }

    missing = sorted(manifest_ids - capability_ids)
    extra = sorted(capability_ids - manifest_ids)

    if missing or extra:
        print("capabilities sync: stale")
        if missing:
            print("missing in capabilities.json:")
            for item in missing:
                print(f"  - {item}")
        if extra:
            print("not backed by manifest.json:")
            for item in extra:
                print(f"  - {item}")
        return 1

    print(f"capabilities sync: ok ({len(manifest_ids)} OpenAPI capabilities)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: domains/knowledge-update/check_capabilities_sync.ts
#!/usr/bin/env node
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = join(dirname(fileURLToPath(import.meta.url)), "..", "..");

function readJson<T>(path: string): T {
  return JSON.parse(readFileSync(join(root, path), "utf8")) as T;
}

function main(): number {
  const manifest = readJson<any>("domains/datafinder-interface/manifest.json");
  const capabilities = readJson<any>("domains/intent-routing/capabilities.json");
  const manifestIds = new Set(
    manifest.endpoints
      .map((ep: any) => ep.capability_id)
      .filter(Boolean)
  );
  const capabilityIds = new Set(
    capabilities.capabilities
      .map((item: any) => item.capability_id)
      .filter((id: string) => id?.startsWith("datafinder.openapi."))
  );

  const missing = [...manifestIds].filter((id) => !capabilityIds.has(id)).sort();
  const extra = [...capabilityIds].filter((id) => !manifestIds.has(id)).sort();
  if (missing.length > 0 || extra.length > 0) {
    console.log("capabilities sync: stale");
    if (missing.length > 0) {
      console.log("missing in capabilities.json:");
      for (const item of missing) console.log(`  - ${item}`);
    }
    if (extra.length > 0) {
      console.log("not backed by manifest.json:");
      for (const item of extra) console.log(`  - ${item}`);
    }
    return 1;
  }

  console.log(`capabilities sync: ok (${manifestIds.size} OpenAPI capabilities)`);
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exitCode = main();
}
